Wounds, Major Changes and Additions

OtherBiscuit opened the pull request:
# Summary
The Wounds system has been on the backburner for a while, and the design document is thorough enough to work with. This is going to change hitpoints and combat drastically, and thus will be tested and tweaked as I work on it.

What Doesn’t Kill You…
The way you “die” is being massively changed. With my very limited medical knowledge, oxygen is the primary reason the body shuts down. This will be game-ified and abstracted heavily, but will affect the way the player is hurt.

Other elements like blood pressure are coming, but will mainly be extensions of current effects like hypovolemia (following the example) and used for symptoms. Most elements will affect oxygen in some way, followed by circulation (hypovolemia > loss of blood pressure > loss of circulation > loss of oxygen in essential organs). This won’t be micromanaged, but will come into play mainly with tourniquets and similar.

Symptoms will assist in diagnosis by the player, as a sort of notification system to help identify major issues.

This also lets writers expand how diseases and injuries work, defining how exactly the player is being hurt beyond “+10 damage”.

How much this expands should be under _a lot_ of scrutiny by the head devs, so I don’t bog down the game or performance. It should ideally be a few core counters, like oxygen is presently.

Effects Replaced
A good deal of effects are being moved to the medical menu, and their effects on the player handled by medical code. You inject Adrenaline (or activate it via your mutations/bionics) and create a tiny wound on your body, with the Adrenaline contaminant, which enters your bloodstream and triggers the effect. Poison and venom are similar.

Assist Innovative Devices
A new pre-cataclysm company which developed AI-assisted medical tools for the homes and the hospitals of New England. Han-DY-alysis (Product DY), SECIL (Product SC), and ERIC (Product ER) will be there to help the player with their more dire ailments.

(A.I.D. products fill the role of a tutorial on more complex medical treatments, going further than Medical Menu tooltips, and providing assistance with treatments. Traditional medical machines are also found more consistently but provide no assistance and require medical knowledge to use. Usually they’re found as vehicle modules on wheels, unable to be removed. SECIL is furniture.)

Han-DY-alysis hosts a (relatively) compact dialysis machine with blood-testing capabilities. Its onboard AI (basically a combined chatbot and medical robot) makes it simple to use for anyone, just connect the machine to your body and let it handle the rest! Han-DY-alysis is commercially available and often found in homes and doctors offices alike—or in hospitals alongside traditional machines of course.

(AI is named “Heba al-Assist”, came up with the name when I was still trying to fit the naming scheme into puns. Like ‘dialysis’, alysis, al-assist, of the family Assist. Thought it might be interesting for characterization, so I kept it. If anybody with more experience in Middle Eastern culture would like to contribute more to her that would be amazing!)

[WIP note: Provides medical tools related to blood, and gives a slight boost to treatment successes. Thinking about adding the potential for user error, if your character makes a mistake while setting everything up. “Oh s#!t I forgot to plug the intake into the machine, my blood is pouring out!”]

SECIL is an evolution of the famous (and recalled) Mr. Stem Cell. Now with the same AI and medical capabilities of his sister, the Stem-cell Enhanced Care for Intensive Liabilities is present in hospitals and medical centers for use by medical professionals. SECIL hosts a wider variety of medical tools and is capable of providing a stem-cell treatment (with appropriate resources).

[WIP note: “SECIL! I NEED YOU SECIL!!”. Jokes aside, SECIL provides almost every medical tool you’d find in an emergency room, greatly boosts your treatment successes, plus the return of Stem Cell Treatment. In the real world, stem cell treatment is only cleared for use in blood-loss, but research has been done on healing wounds and treating ordinarily untreatable conditions which involve currently irreparable cells. 

If we ever make nerve damage more common, or add brain damage or spinal-damage paralysis—both of which I say are terrible ideas—this could be a good option.]

ERIC is an experimental compact AI-assistant installed on a few select ambulances for trials in the field. Emergency Response Intervention - Compact is equipped with medical tools for trauma which require emergency medical services, designed to assist first responders. ERIC was developed quickly as a response to the rioting and chaos leading up to the Cataclysm, and was never refined or distributed—even within New England.

[WIP note: Found with ambulances, slotted in where another module would be similar to a stretcher. Thinking about cutting this and replacing it with another assistant. Where Han-DY-alysis provides information and help with contaminants and blood loss, ERIC helps with trauma in the field that the player would likely encounter. It’s meant to be a step close to Han-DY-alysis, but short of SECIL.

A separate Assist product to take care of this found in similar circumstances might be better.]

[General WIP notes: An Assist product for fire departments focused on burns (acid or otherwise) would be good. Not sure how necessary it may be, but I don’t think it should be relegated to SECIL, especially with how dangerous acid is becoming.]


I’m using a lot of ideas from various sources, all of which will be attributed.
-Original Wound Design Doc (Issue #64424)
KorGgenT, I-am-Erk, Shiubiha, Venera3 and zachary-kaelan contributed the framework for the implementation of wounds and contamination.
-"Flinch", a mechanic for making speed penalties less deadly and annoying (Issue #73050)
I-am-Erk laid out a mechanic for flinching, rather than pain universally slowing the player. It’s been waiting for the wounds system for a while.

## Combat Changes
These are WIP and will need to be extensively tested and changed.

If things across the board are far too deadly, the JSON values for injury depth and thresholds can be adjusted (eg. “Small Cut” will only become a different wound at 20 damage rather than 10)

If a specific encounter is deadlier than intended, the damage values of that specific monster/attack can be adjusted (eg. A zombie variant can cause extremely deep wounds on specific vitals, only countered by armour not available at this point in the game)

Pain changes, limb loss WIP, new Exodii merchant for prosthetics, repairs and regrowth, all are WIP

### Bifurcation, Decapitation, Etc
Without insane mutations (more than likely restricted to the undead or hostile mutants) this would definitely be fatal, or at the very least put an end to your adventuring without exotic repair and replacement.

However, seeing (or even being) someone get their head chopped off, and watching crab legs sprout from the stump as it scurries off like The Thing, would be freaky as all hell.

# Purpose of change
As great as Hitpoints are, they’ve had their issues in the past with balance and “realism”---our ever-present lofty ideal. In the real world, stubbing your toe a hundred times doesn’t break your femur; and this isn’t fun in-game either.

This will also make changes to pain and the player’s body, as combat changes made by wounds will require rebalancing and the associated changes along with it.

# Describe the solution
## Wounds:

Rather than hitpoints, your main limbs will accumulate wounds.

When attacked, you will accumulate wounds. If these wounds overlap, they are merged, increasing the surface area and potentially severity. Each wound is affected by its depth, type, and surface area.

This is done by a “layers” system, from your skin to your bones to the skin on the other side. Attributes affect the resilience of these layers, similar to the interaction between armour and hitpoints as it is now.

For example: Bones are pretty damn strong, slicing through them isn’t easy by normal means. However, having weakened bones from some preexisting condition or deficiency might make your bones a bit less resistant to damage.

On the more extreme end, having hollow bones from mutation will make your bones much easier to shatter—exposing you to more unique wounds due to your new biology. While hollow bones may have the same material and strength, they aren’t solid, and you’re at risk for injuries that wouldn’t happen with solid bones (like getting shards of bone stuck inside your healing bone)

### **Layers**:
As recommended by IdleSol, the body is made up of layers.

[Layer 0] Anything outside your body, eg armour and clothing. This is just your clothing layers.
[**Sub-Clothing**]
Damage absorbed by your outer layers will still affect what’s under it. Your plate carrier stopping a rifle round is fantastic, the bruises and broken ribs aren't.
Different parts of the body hold different vitals. Say ~3% of the surface area of your leg is above a vital area, if an attack hits this area, it will affect your vitals. Say your femoral artery is after the muscle (Layer 3) but before the bone (Layer 4). If damage is deep enough to affect Layer 3, on this 3% area, you’ll have a chance of striking the vital. This is increased if the second threshold Layer 4 is reached.
Bionics will have a defined depth layer and coverage, acting similar to vitals
[Layer 1] Hair/Wool/Etc, this is your main source of cushioning (reducing blunt force from attacks stopped by your other layers)
[Layer 2] Skin, or equivalent. 
[Layer 3] Tissue, muscle, etc
[Layer 4] Bone
[Layer 5] Deepest tissue and organs
[**Other Side**]
Wounds deep enough will reach the other side of your body.
[Layer 6] Other Side’s Bone
[Layer 7] Other Side’s Tissue
[Layer 8] Other Side’s Skin
[Layer 9] Other Side’s Hair
[**Clothing**]
[Layer 10] Other side’s outer area

### Example
For some examples of what wounds will look like:

#### **ID**:
The source of the wound affects its behaviour, healing method, etc. A third degree burn is much different than a burn from acidic material. Various injuries are assigned an ID in JSON, with the information for their healing and attributes. This makes it easier to create new wounds from unique sources (even if just a copy-from variant)

#### **Depth**:
The deeper the wound, the closer it is to your major blood vessels, organs, bones, and vital areas. A small through-and-through wound that doesn’t hit your vitals is lucky and fantastic, and certainly may happen. Just the same, you might be struck in the thigh wrong and burst your femoral artery.

In-game, this will have to be balanced extremely carefully.

##### **Bionics and Mutations**:
Additionally, your mutations and bionics may be in the way of different attacks and such. 

A subdermal armour bionic situated just below your skin is designed to take and slow down the force of attacks, and won’t do anything too terrible if it breaks (relatively speaking. Only a small hole or crack in the armour that sprayed particles of material into your body, like minor shrapnel)

Conversely, an arm shotgun may have its hammer knocked out of place, or the deployment mechanism gets jammed, causing it to not function. On the higher end, a portion of your cloaking system bionic being damaged may send sparks into your body or uncontrollably power on/off occasionally. Or even your time dilation bionic occasionally malfunctioning and making parts of your body react slower/faster, causing friction and frying your reaction time and nerves.

For mutations, having a huge rack (ANTLERS, damn it! Get your mind out of the gutter!) would add surface area on your head composed of only a single layer: bone, slightly different to your other bones. Different resistances and less painful to break and grow back. Unless of course these antlers are just growths of normal human bone, but that’s not relevant here.

This also gives a place for Issue #79031; getting half your antler knocked off and having it grow back means losing an ear/mutated ear would have it heal over in some way—still recognizable as a scar or mutated clump of flesh. The body horror of your system healing towards your mutation rather than your body is interesting; you’ve changed irreparably (unless you find some purifier…)

#### **Surface Area**:
An abrasion across your forearm would accumulate dirt and filth faster than a cut on your knee, and would need more material to cover. Burns are extremely dangerous for this reason, as they are in real life.

## Contamination
Having foreign substances in your open injuries is bad. Depending on what causes the wounds, different contaminants will find their way into your wounds, causing different infections. In general, uncovered wounds accumulate generic "filth", causing a variety of your mundane, earthly infections and diseases.

-Contaminants drain into your system, depending on their "spread factor". This doesn't cause the contamination to disappear, this only delays how quickly contamination sets in. If the system contamination level reaches its threshold in the wound, the wound becomes infected with the specified infection.
-Infections have a "severity", similar to effects having multiple levels. Deeper wounds will be more likely to begin more severe, especially if close to vitals. Some infections may have a single level, others may get worse over time unless combated, some progress regardless.
-Infection makes healing much harder, sometimes outright

This means older effects getting a revamp; take tetanus for example:

1. Initial contamination. For filth, this is caused by leaving a wound uncovered in general, and starts higher if the wound is caused by something like mounting a broken window. For rust monsters, tetanus is the only infection caused by the specific contaminant
2. Contaminants in the wound are left uncleaned and drain into the system. If more contaminants of the same type enter the wound, and the system contamination for this wound crosses the threshold
3. The threshold is crossed, and a delay is started. At the end of the delay, a percentage chance is calculated by the difference in the contamination value and its upper limit. When that delay is reached, the chance is rolled and the delay begins again **as long as there is still contamination in your system**
(ie, System Contamination has a risk of infection at 10, and will always infect at 110. 20 contamination enters your system before you wash your wound. 10 points past the initial limit _out of_ 100 points between risk and guaranteed infection. 10% chance.)
4. The system contamination constantly lowers based on its "decay", which represents your system trying to filter out and defeat the infection. When the infection chance is rolled, if the infection strikes, system contamination is reduced and the infection severity increases (if you are infected currently)
5. Initial severity is affected by depth, and infection chance is affected by depth.

EQUATION WIP: Calculations will be added
Contamination > System Contamination > Delay > Roll based on factors > Roll for infection type (if multiple) > Repeat until no system contamination present

As far as I know, tetanus has historically only triggered from mounting broken windows or rusty objects, more recently through certain monsters. In the real world, tetanus is a bacterial infection caused by a bacterium commonly found in soil worldwide.

Tetanus will be one of the infections possible when you leave your wounds uncovered and accumulating the filth contaminant.

### Notes on Contamination
Filth will be the generically used contaminant for general contamination while out and about, Zombie Filth will be the specific contaminant which causes "infection" as it currently is. Where filth has a chance of causing this infection (due to XE-037 presence worldwide), Zombie Filth specifically causes "infection", and accumulates from attacks by zombie monsters.

WIP

## Describe alternatives you've considered

### **Misc Notes**

#### **Bionics and Mutations**
Bionics may be changed or JSON-ified, moving the specific changes to a generic damage function for future use may be more reasonable.

Medical and Chimera mutations are getting huge bonuses. Ultra-fast healing and morphing can make lethal wounds a minor setback, and give you effective recovery methods mid/post-combat. Bleeding isn’t a problem when your wounds are instantly covered with reactive bone growths and clots form in seconds.

Lizards and such get this to a much lesser degree, but limb regrowth is nothing to scoff at.

#### **Organ Damage**
A collapsed lung or a punctured heart would be invariably fatal to the average survivor. Not to mention brain damage from being knocked around. I assume instant deaths and common permanent debuffs are not intended, barring extreme circumstances. Unless I'm given a greenlight for some makeshift medical equipment similar to what you'd find in a staffed ambulance, the initial implementation will only trigger insta-death if the heart, lungs, or brain sustain immense damage. Other vitals and organs can be added with long-term damage changes.

#### **Sticks and Stones**
In the real world, XE-037 healing factor or not, you're not repairing a shattered jaw on your own, nor can you reasonably recover with a broken spine. We can explain away healing as XE-037 kicking into high gear, but it still has a limit. For the initial implementation I'm limiting damage to "cracked" bones, a step below breaking or shattering. A cracked jaw might make eating painful and difficult, but it can be reasonably handled on your own; later on, we can figure out a solution attainable through challenge or faction debts (maybe even incorporating the "super-soldier serums")

#### **Assist Innovative Devices**
This is a huge new corporation with new lore and might not be the best approach for their purpose, or might not fit with lore. They aren’t very involved with top secret organizations like XEDRA, maybe at most giving them first-dibs on ERIC or similar technology.

They’re meant to be the mundane step below Super Soldier Serums, only guaranteeing you can recover on your own.

#### **Moddability**

##### Old HP System
This is meant to be heavily moddable and compatible. So what if your mod hinges on using the old HP system?

I’m including a generic “damage” wound which fills this purpose, and will work as the old hp system does, even excluding organs and arteries and such. This doesn’t catch everything, but I’ll do my best and will remain very open to the concerns of modders.

WIP

### Testing
WIP

# Additional context

### Contaminants System
This is **extremely rough** and **convoluted**, any suggestions are very welcome!

### Suggestions...
...are **extremely welcome** and **appreciated**! This is all rough, and I'm putting together multiple ideas from the design document and scenarios brought up previously. Credit will be given to those whose ideas were used to build this system.

Suggestions from this PR:

"Layers system for body, giving resistance for each layer individually and helping with calculations" _HIGHLY ABBREVIATED_
-IdleSol

_This is genius, and can solve the issue of mutations/bionics altering resistances; simply applying armour bonuses to certain damage sources on specific layers_

“Resetting bones and certain injuries, losing teeth, and similar” _HIGHLY ABBREVIATED_
-ADekema

_Like cauterization, resetting bones will be_  a  _option for injury treatment, but not without drawbacks and risks. And with the recent additions of a dentistry office, I could definitely add a medical machine found there and the needed items. But I don’t think it would be needed in general, and would bog down gameplay a bit without much effect outside niche problems._

WIP

OtherBiscuit committed:
Initial Migration of KorGgenT's work pt 1

OtherBiscuit committed:
Initial Migration of KorGgenT's work pt 1

OtherBiscuit commented:
Oh Christ I forgot to move this over to my branch. Very sorry.

OtherBiscuit commented on src/bionics.cpp:
```diff
@@ -1785,7 +1785,7 @@ void Character::bionics_uninstall_failure( int difficulty, int success, float ad
                     }
                     bp_hurt.emplace( bp->main_part );
-                    apply_damage( this, bp, rng( 5, 10 ), true );
+                    apply_damage( this, bp, damage_instance( damage_type::CUT, rng( 5, 10 ) ), true );
```
Bionics changes may be moot by now, I have to review it. Any extra changes will be caught on the next pass.

GuardianDll commented:
for the future, if you plan to base your work off the Korg commits, you should preserve the authorship of commits for proper attribution

fungamer2-2 commented on src/wound.cpp:
```diff
@@ -0,0 +1,226 @@
+bool wound::overlaps( int target ) const
+{
+    if( location + id->size <= 100 ) {
+        return location >= target && location + id->size <= target;
```
```suggestion
        return location <= target && location + id->size >= target;
```

Just fixing the `wound::overlaps` method

OtherBiscuit replied:
Thanks a bunch! Gonna handle it with the next pass

fungamer2-2 commented on src/wound.cpp:
```diff
@@ -0,0 +1,226 @@
+    if( location + id->size <= 100 ) {
+        return location >= target && location + id->size <= target;
+    } else {
+        return location >= target || location + id->size - 100 <= target;
```
```suggestion
        return location <= target || location + id->size - 100 >= target;
```

RenechCDDA commented:
Just to be clear this absolutely does need proper attribution if you want it be mergeable.

ADekema commented:
How will this interact with HP mutations or with differant survivors having differant HP amounts based on weight and height?

OtherBiscuit commented:
> How will this interact with HP mutations or with differant survivors having differant HP amounts based on weight and height?

HP mutations will change to make the survivor more susceptible to wounds (more than likely, just slight changes like small negative armour values to certain damage types unless a better idea comes along)

Hitpoints are mainly replaced by blood volume, which would increase with your size. Lose too much, you'll go into shock (and anemia), and eventually not have enough to keep your body oxygenated.

IdleSol commented:
<details><summary>offtop</summary>
If each limb could be thought of as a set of layers with different coverage. And then look at which layer the damage reached, then you could adjust the depth of the wounds and their degree.

Conditional:
- L1 = hair/wool (100%)
- L2 = skin (99%)
- L3 = meat/muscles (90%)
- L4 = bone (70%)
- L5 = internal organs (0..50%)
- L6 = bone (70%)
- L7 = meat/muscles (90%)
- L8 = skin (99%)
- L9 = hair/wool (100%)

Layers have their own armor and HP. HP is needed to reduce armor and adjust the varying degrees of healing speed.

The cube determines which layers are affected. And the damage and its type determines which layer the damage will reach.

(It's similar to clothing. But the die is rolled once, not for each layer. And the die doesn't determine the hit. It determines exactly where you are hit, assuming you are hit.)

A few examples:
1. Cube 100. Damage will hurt the L1 and L9 layers. A little damage to the wool.
2. Cube 70. Damage will hit all layers except L5.
- The damage value is enough to reach L9. You've been stabbed through. There's a lot of blood, bones are broken, muscles are severed. But no internal organs were hit. (L5)
- the damage value is enough to get you to L4. The situation is similar to the first, but less dangerous. You've been cut to the bone, but not stabbed through.
- the damage value is enough to reach L2. Your skin has been slightly scratched.
</details>

KheaHyena commented:
Ooh, really interesting, thanks for picking it up!

I am particularly curious in how will this affect certain attacks and abilities (Flesh-Raptor comes to mind as a possible earlier enemy) that have the impale strike, since it would mean these could be **_extremely_** deadly when Vitals are involved. Also on how will regular armor be more relevant to prevent deeper wounds from reaching said Vitals.
The heightened natural regeneration we have from exposure to XE037, is also something I am interested on how will it be tackled.

Come to think of it, it also raises other curiosities with Health Care, like stitching wounds and such...

OtherBiscuit commented:
> I am particularly curious in how will this affect certain attacks and abilities (Flesh-Raptor comes to mind as a possible earlier enemy) that have the impale strike, since it would mean these could be **_extremely_** deadly when Vitals are involved. Also on how will regular armor be more relevant to prevent deeper wounds from reaching said Vitals. The heightened natural regeneration we have from exposure to XE037, is also something I am interested on how will it be tackled.
> 
> Come to think of it, it also raises other curiosities with Health Care, like stitching wounds and such...

Healthcare is something touched on a little bit later, almost done with the description up there. Most enemies won't intentionally target vital areas outside of specialized attacks (CUT_THROAT would target your neck, trying to inflict a deep slash wound). Getting unlucky with a strike can definitely be deadly, but I'm not intending to make it a run-ender.

For your example, perhaps enemies get stuck in your wound, this might make it filthier but it also stops your bleeding for a moment until the foreign object is removed. The attacker jostling in your wound would be damaging of course, and also insane for body horror. Huh, neat idea.

Either way, something like that would be important to balance, this isn't a case of "Oh, you should've been armoured, you should play my way". Realistic or not, you've got options irl

KheaHyena commented:
> Either way, something like that would be important to balance, this isn't a case of "Oh, you should've been armoured, you should play my way". Realistic or not, you've got options irl

A very thoughtful way of managing the Combat system for sure.
I am thrilled to see how it will all pop out in the end. The very notion that combat will be more than just managing a Health Pool and using it like a resource, but rather having the need to put more thought into it with all the problems that can arise (and necessary healing afterwards) makes this very interesting in several ways besides gameplay itself. I'm personally very callous when diving into Melee Combat when I'm mildly prepared due to it's simplicity, which, granted, has been worsened by the extension of the Grabbing mechanics, which are already deadly in certain circumstances.

One effect I am eager for which will undoubtedly happen (unless one is extremely cautious) is permanent damage (broken spines and jaws as mentioned, or even tissue scaring). For long term characters this will mean they will require to go through either the process of CBMs or Mutations, to compensate, mitigate or remove their injuries. It's a cool notion, to forcibly adapt one's Character to the ever changing and doomed World they now live on.

Speaking of body horror, my goodness, just imagine the Dermatik spawn.
I guess there is also venom, suffocation, smoke and poison inhalation to think about, but those seem a little less to deal with than the entirety of the damage system (not that our lungs would agree).

KorGgenT committed:
KorGgenT Migration pt 2

TODO: Need to check over some stuff that's been JSON-ified, and to do another check for extra instances of hardcoded damage that's been added.

Character Modifier Tests: Removed the tests for wounded limbs, presumably because wounds would alter the way this is calculated. Need to replace.

Co-Authored-By: OtherBiscuit <136829014+otherbiscuit@users.noreply.github.com>

OtherBiscuit committed:
KorGgenT Migration pt 2

TODO: Need to check over some stuff that's been JSON-ified, and to do another check for extra instances of hardcoded damage that's been added.

Character Modifier Tests: Removed the tests for wounded limbs, presumably because wounds would alter the way this is calculated. Need to replace.

Co-Authored-By: OtherBiscuit <136829014+otherbiscuit@users.noreply.github.com>

KorGgenT committed:
KorGgenT Migration pt 2

TODO: Need to check over some stuff that's been JSON-ified, and to do another check for extra instances of hardcoded damage that's been added.

Character Modifier Tests: Removed the tests for wounded limbs, presumably because wounds would alter the way this is calculated. Need to replace.

Co-Authored-By: OtherBiscuit <136829014+otherbiscuit@users.noreply.github.com>

OtherBiscuit committed:
Merge branch 'master' of https://github.com/OtherBiscuit/Cataclysm-DDA

OtherBiscuit commented:
Re-adding the proper author attribution, Git Bash makes my head hurt, please mind the mess and hold for fixes

OtherBiscuit committed:
Merge branch 'master' of https://github.com/OtherBiscuit/Cataclysm-DDA

OtherBiscuit commented:
> Re-adding the proper author attribution, Git Bash makes my head hurt, please mind the mess and hold for fixes

Think I've got it fixed, please ping me for any errors I've made.

Nebnis commented:
Really looking forward this PR because i love biology stuff. Hopefully there will be infrastructure for unique wounds down the line like exotic attacks from nether beings,ex: A wound that takes 10 times longer to heal or one that slowly eat away your bones until it finally breaks.

ADekema commented:
I do like this as it effetively gives the player weakpoints that a monster might target. Some animals instinctively go towards certain weakpoints. This could make things like dogs and thier zombie counterparts far more dangerous as they still instinctively go towards the most deadly area´s to hit. I don´t think the majority of enemies should have this (Giving this to amalgamations would be a silly idea). We might also want to look at the neck slash attack as a result as it might inadvertantly give certain creatures a massive spike in danger. 

I also dislike the idea of permanent damage to the same degree as we have in real life. I would personally like to see some kind of compromise where the extra regeneration that humans in DDA have also translate to a higher regenerative capacity. Not to the degree of regrowing limbs but to the degree that even stuff like scars and connective tissue´s eventually heal fully if the proper treatments are applied. So for scars you would need to be able to preform deep tissue massages and stretches to break the scars up before the fully dissapear and some injuries like fully severed tendons might require surgery to reconnect but would eventually go back to normal over a course of weeks to months (So still a very long time in game terms.)

So if you really mess up and start to outright lose bodyparts or get it mangled the damage is permanent. But if it isn´t you can probably make a full recovery with perhaps some surgeries from a friendly faction or two. If the damage is permanent than your only option would be to either get bionics (agianst hefty payment) or start gambling with mutations.

GuardianDll commented:
What's the current status of this PR?

OtherBiscuit commented:
> What's the current status of this PR?

Marked the attributions, waiting for KorGgenT to be available to confirm a couple things about their code, and currently working on replacing instances of hardcoded damage with the new format. 

This as a whole is going to take a while, I can move everything over to another branch if it would help?

GuardianDll commented:
I'm asking because i didn't track this PR, and i have very high hopes for it

OtherBiscuit commented:
> I'm asking because i didn't track this PR, and i have very high hopes for it

I do too, I’m still working and I’m hoping it does its job well.

Thinking about taking this to the general public and asking their opinions once I’ve finished the description, would that be okay?

GuardianDll commented:
Yeah just post how the json of the result would look like, and how it would interact, and that would be good

harakka commented:
KorGgenT isn't currently active so you likely need to get those answers from someone else. Is it something that's actually stopping you from proceeding?

OtherBiscuit commented:
> KorGgenT isn't currently active so you likely need to get those answers from someone else. Is it something that's actually stopping you from proceeding?

Just wrapping up an updated design plan to run by the community and devteam, hope KorGgenT is alright

John-Candlebury commented:
Hey I really dont have the time to parse how the back end of this works, I also donts see any json defined wouds, so I'll just want to ask:

- Given that all damage instances and armor values are plain numbers, how do they actually relate to wounds? 
- Does the player limbs have an abstract HP value still, which is decreased by applied wounds?
- How do you plan to make wounds json driven? Please consider that mods will need to blacklist or disable vanilla wounds.
- Is it possible for mods to still simulate HP with this system? Such a simulation would basically looks like: The player receives one "1hp wound" for every single point of damage received, limbs are disabled when they stack X number of these wounds, these wounds resolve overtime or with simple treatment and all other wounds are disabled.

Imfairly concerned about this because wounds dont exactly fit Aftershock's design paradigm, and I just spent the past months crunching numbers to get the mods balance in a nice carefully calculated place with: #79762 and all the other related prs.

It was my understatement that the original design conception of wounds was thet they were basically persistent and compounding sources of debuffs and max_hp reductions, but what I'm seeing here seems to be the dwarf fortress randomness based  system of "a strong attack either harms you very little or decapitates you outright".  As one of the main json content creators for this game, this concerns me deeply because: Its not what was originally planned, and this is effectively impossible to balance consistently.

OtherBiscuit commented:
> Hey I really dont have the time to parse how the back end of this works, I also don't see any json defined wouds, so I'll just want to ask:

This is all a result of the messiness of this document, I'm very sorry for the lack of transparency and clarity with certain pieces. I'm still formatting the document for readability for a semi-final summary.

> Given that all damage instances and armor values are plain numbers, how do they actually relate to wounds?

Wounds are similar to the way effects are currently handled, a threshold for becoming a different type and a definition for what would cause it. 
For example, simplifying other attributes:
You are hit across your arm and leg for 15 slashing damage. 
Your arms are protected, and the final damage dealt is converted to bashing, ~4 Bashing damage. "Minor Bruise" wounds form from bashing damage anywhere from 1 to 10, so the area of your arm struck gains a Minor Bruise.
Your legs aren't protected enough to dampen the damage, 15 Slashing damage done in its entirety. "Cut" wounds form with slashing damage from 1 to 10, "Gash" wounds form from slashing damage 11 to 25; so the area struck gains a Gash.

Combining damage types would result in a similar wound to those caused by both singular types. This is a rough explanation, I should have the more polished and easier to understand bits soon.

> Does the player limbs have an abstract HP value still, which is decreased by applied wounds?

No, a broken limb as it exists currently would be equivalent to a ~mid-level bone break from bashing damage; not cut through, not shattered, but beyond a crack. 

> How do you plan to make wounds json driven? Please consider that mods will need to blacklist or disable vanilla wounds.

I won't be hardcoding the wounds, nor the applications. It'll be similar to replacing items and their groups. Replacing a generic "Deep Bruise" with something custom for a mod would be as simple as it would for replacing any other item.

> Is it possible for mods to still simulate HP with this system? Such a simulation would basically looks like: The player receives one "1hp wound" for every single point of damage received, limbs are disabled when they stack X number of these wounds, these wounds resolve overtime or with simple treatment and all other wounds are disabled.

For something as you describe, absolutely! Redefining the available wounds with a generic "Damage" wound, blanking the defined layer resistances (thus only armour would reduce damage, though bash damage from blocking force would still apply), and causing it to change to a "Broken Limb" wound at a certain threshold. And of course defining the recovery time to be unmodified by other variables, and defining the recovery requirements as None.

However, I'd need to alter the implementation a bit to make sure the wounds would be tracked over the entire limb/body, add capability to cause effects (like killing the player) upon a wound being applied (could also be useful for certain unique attacks), and the like. These options could also be useful for other core-game wounds.

All of this is extremely important for mod support, so thank you for bringing it to my attention, I hadn't considered how badly this may mess with existing/future mods. I planned for the damage converting to wounds to make compatibility as non-invasive as I could, but this needs much more TLC and features to make sure authors have the tools they need.

> Fairly concerned about this because wounds don't fit Aftershock's design paradigm, and I just spent the past months crunching numbers to get the mods balance in a nice carefully calculated place with: https://github.com/CleverRaven/Cataclysm-DDA/pull/79762 and all the other related prs.

Oof, yeah, balance is going to need to be managed a lot before I set this loose. Trying to keep this in line with the initial ideas for wounds, being inconsequential at first but spiraling out of control if not managed properly; through this, I'm pretty sure things should be fine damage-wise for the most part, the only major shakeup is bashing damage from armour blocking and vulnerability to blood loss and loss-of-movement.

I do NOT want to add extra work for anyone else, and I want to do my best to handle anything I break myself to not drag anybody else down.

The simplest solution for mods which would have certain encounters/specific attacks suddenly become much more deadly than intended, would be to increase the values on wound thresholds/tweak damage. It's a duct-tape fix, and it needs much more work to make sure I don't trample other peoples' visions for the base game and their own mods.

> It was also my understatement that the original design conception of wounds was that they were basically persistent and compounding sources of debuffs and max_hp reductions, but what I'm seeing here seems to be the dwarf fortress randomness based system

Yes and no; picking on the low-hanging fruit for an example: Pain and Speed. One or two bad wounds compounding and forcing you to retreat is intended, but this needs to not feel like BS. Tendons in your hand getting sliced would hinder your hand's manipulation and accuracy, getting your Achilles' sliced (eugh) would seriously hinder your movement, etc. Injuring muscle groups and major points like your elbows and shoulders would do the main job of debuffs.

HP-reduction would be somewhat covered by certain vulnerabilities caused by other wounds. Getting hit on a bruise would spike your pain, hit on a clotting wound would make it bleed again, but given that HP isn't really applicable here the equivalent is lowering resistance of certain layers due to certain wounds.

And the randomness aspect is an extremely delicate beast, but it won't be a dwarf fortress-esque "the chances of your skull getting caved in by a stray punch are lowered by wearing a helmet, but that chance is still very present".

Moreso "The areas with major arteries underneath over your entire body are low, and under other layers, so even an unlucky roll still has a decent chance to not open your carotid just because a zombie child nicked you"

Again, balance is going to be key for base-game, but mods might be screwed up by these changes. Making the system more customizable is going to be near the top of my to-do list.

> As one of the main json content creators for this game, this concerns me deeply because: Its not what was originally planned, and this is effectively impossible to balance consistently

There were many ideas for what could be used, constantly being added onto and introduced. I compiled what I thought could be used _while_ fitting in the initial plan as much as possible. 

Making the system interesting, useful for interesting encounters, simple enough for players at any experience level, and not micro-manage-y is a serious balancing act; and that's without accounting for future ideas and mods. My worst fear is reigning in creativity and preventing people from contributing brilliant ideas.

Quality of life, modability, all of it is being considered.

And of course, the proper examples of JSON implementation would answer many of these questions, and is probably what you need. I'm very sorry they aren't ready yet, not trying to drag this out for attention or to prevent anybody from working, this is just a lot to comb through and balance.

ADekema commented:
Two question and a suggestion:

How exactly will damage blocked by armour be converted into bash damage? For ballistic damage this is easy as they are based on kinetic energy but for cutting or piecing damage this get´s harder. Many weapons or enemies have higher damage number to account for things like sharper cutting surfaces or points. But the cutting surface doesn´t matter squat if it can´t get through your armour. So a razor sharp katana might do more damage than a blunt porely made sword of the same weight but if I hit a metal breastplate with it the katana would still do more bash damage on account of having higher cutting damage?

Will more exotic damage types like fire, acid and electricity be able to kill you in the initial implementation? As you currently plan it only direct severe trauma to the brain, heart or longs can kill a player. But what if you get severe burn covering most of your body? Do you somehow survive that? What about poisons?  Is there a threshhold where we just say ¨80% of your body is covered in burns you die now.¨? Is there some kind of internal counter that keeps track of biological damage and after a certain point says ¨You are to damaged on a cellular level to keep functioning and drop dead.¨?

I think that for the initial implementation it would be good to introduce a few wounds that are more severe than a cracked bone. This would get players used to the idea of some wounds being extremely detrimental and needing others to help repair them. Some examples I can think of:

1. Broken bones. Not outright shattering as that would require surgery but simply a bone that has a fracture that requires a splint to repair. Probably requires your character to realign their bones by hand (Extremely painful) and than immobalize them with a splint or a makeshift cast. This is already the way we treat broken limb currently so I don´t think this would be a very good wound to set the limit at as we already have the means to treat it with splint and most people are familiar with it. The only thing that needs to be changed is splints for smaller digets (fingers, hand, foot) and you need to add a huge pain penalty for putting on the splint as the process also involves realigning your bones.

2. losing teeth. Treated via dentist with either quickly reatatching knocked out teeth or getting crowns. would vanish if you get a tooth mutation and a new set of teeth. PRobably need to implement mouthguard before implementing this and figure out a way to have and armour piece that is located after the skin layer.

3. Broken/shattered nose. Can be realigned yourself if your quick enough but would probably hurt a lot.

OtherBiscuit commented:
> How exactly will damage blocked by armour be converted into bash damage?...
The damage types' impact will be similar, a % reduction based on damage and armour values and converted to bashing. Ballistic damage (which is to my knowledge separate from piercing) would be less reduced, and would lead to wounds similar to high end bruising/bashing injuries. I'll need to do some research on how these wounds actually work IRL 

> Will more exotic damage types like fire, acid and electricity be able to kill you in the initial implementation?
I'm still working on how deeply these systems will go, but I plan on adding various values (potentially moving some vitamin counters, whatever would be most applicable according to other devs) for things like blood pressure. Again, I need to research how these wounds work with the body IRL, and just how much I can handwave in the name of balance with the blob's interactions with the body.

Poisons will be contaminants in the system, as mentioned in the section above. A venomous bite would just leave extremely fast-absorbing contaminants.

I don't want to exaggerate too much, but the player will be much more durable than normal. With your example of severe burns over much of the body, the player could pull a Joshua Graham with proper medical treatment, for lack of a better example. Again, I need to do more research.

> I think that for the initial implementation it would be good to introduce a few wounds that are more severe than a cracked bone...
Oh absolutely, the stuff about immediate lethality only being triggered from damage to certain organs is a temporary solution. More than likely the route taken will be that organs and such will be much more easy to regenerate than they would be in the real world, but extreme damage would still be very dangerous. Rough makeshift solutions would be available, but only in circumstances they'd make sense

I've been tinkering with the idea of re-adding cauterization as an option through the medical menu, but only even remotely usable with ridiculous First Aid skill. Similar treatments are ill-advised for those without experience, and still potentially dangerous even with experience. Desperate times may still call for desperate measures...

> Edit: in line with my second question I would like to ask if damage being converted into bash damage by armour will also be applicable to monsters?
Logically yes, but that falls under combat balancing. I feel it's the same boat as tracking things individual limb hp for monsters; too much for too little, and upsetting balance.

OtherBiscuit commented:
Almost upon the second month of this PR, still working, the next edit will likely be the final version of the PR description. From there, if the devs have any additional ideas or concerns, they'll be added and tweaked.

Thank you all very much for your patience, this is a huge undertaking, but splitting the PRs would risk causing balance issues and bugs between updates. Keep asking questions and making suggestions! Sorry this is becoming as long as the original design documents' discourse.

OtherBiscuit commented:
Updated, still working, please feel free to add any ideas!

Currently wrapping up the hardcoded stuff, going to move to limbification and limbs, then the last of the core mechanics. Then I'll start on the JSON, added machines and items, balancing, etc.

OtherBiscuit commented:
It didn't carry over the formatting. And I'm considering closing this and splintering to a branch to make collaboration and documenting more convenient.

OtherBiscuit commented:
Proper Update: Closing this soon, branching into existing Limb Mod for Limbification, then moving onto the above changes to avoid dumping so many changes.

I knew this is what I should’ve done from the beginning, and what was recommended. Sorry for the confusion.

I am not giving up however! This is going to be finished, come hell or highwater.

OtherBiscuit commented:
Following a few months, I realized I'm reinventing the wheel. This is an overcomplicated version of what Rimworld does. Huh.

Drew4484 commented:
> Following a few months, I realized I'm reinventing the wheel. This is an overcomplicated version of what Rimworld does. Huh.

Complicated doesn't make it a bad thing, especially if the complicated stuff is happening in the background. If you still want to make changes along these lines, feel free to ask here or in the Devcord for review, feedback, or rubber ducking.

Nebnis commented:
> Proper Update: Closing this soon, branching into existing Limb Mod for Limbification, then moving onto the above changes to avoid dumping so many changes.
> 
> I knew this is what I should’ve done from the beginning, and what was recommended. Sorry for the confusion.
> 
> I am not giving up however! This is going to be finished, come hell or highwater.

Thank you for your dedication, this is one of the changes to the game i'm most looking forward
